src/utils/block.ts

```diff
@@ -0,0 +1,229 @@
+  }
+
+  _registerEvents(eventBus: EventBus<Events>) {
+    eventBus.on(Block.EVENTS.INIT, this.init.bind(this));
+    eventBus.on(Block.EVENTS.FLOW_CDM, this._componentDidMount.bind(this));
+    eventBus.on(Block.EVENTS.FLOW_CDU, this._componentDidUpdate.bind(this));
+    eventBus.on(Block.EVENTS.FLOW_RENDER, this._render.bind(this));
+  }
+
+  protected getStateFromProps(props: any): void {
+    this.state = {};
+  }
+
+  dispatchComponentDidMount() {
+    this.eventBus().emit(Block.EVENTS.FLOW_CDM);
+  }
+
+  init() {
+    // this._createResources();
+    this.eventBus().emit(Block.EVENTS.FLOW_RENDER, this.props);
+  }
+
+  _componentDidMount(props: P) {
+    this.componentDidMount(props);
+  }
+
+  componentDidMount(props: P) {
+  }
+
+  _componentDidUpdate(oldProps: P, newProps: P) {
+    const response = this.componentDidUpdate(oldProps, newProps);
+    if (!response) {
+      return;
+    }
+    this._render();
+  }
+
+  componentDidUpdate(oldProps: P, newProps: P) {
+    return true;
+  }
+
+  setProps = (nextProps: P) => {
+    if (!nextProps) {
+      return;
+    }
+
+    Object.assign(this.props, nextProps);
+  };
+
+  setState = (nextState: any) => {
+    if (!nextState) {
+      return;
+    }
+
+    Object.assign(this.state, nextState);
+  };
+
+  get element() {
+    return this._element;
+  }
+
+  getPropsAndChildren = (propsAndChildren: any) => {
+    const props: any = {};
+    const children: any = {};
+    Object.entries(propsAndChildren).map(([key, value]) => {
+      if (value instanceof Block) {
+        children[key] = value;
+      } else if (Array.isArray(value) && value.every(v => v instanceof Block)) {
+        props[key] = value;
+      } else {
+        props[key] = value;
+      }
+    })
+    return { props, children }
+  }
+
+  _render() {
+    const fragment = this.render();
+    const newElement = fragment.firstElementChild! as HTMLElement;
+    if (this._element) {
+      this._element.replaceWith(newElement);
+      this._removeEvents();
+    }
+    this._element = newElement;
+    this._addEvents();
+  }
+
+  protected render(): DocumentFragment {
+    return new DocumentFragment
+  };
+
+  getContent(): HTMLElement {
+    if (this.element?.parentNode?.nodeType === Node.DOCUMENT_FRAGMENT_NODE) {
+      setTimeout(() => {
+        if (this.element?.parentNode?.nodeType !== Node.DOCUMENT_FRAGMENT_NODE) {
+          this.eventBus().emit(Block.EVENTS.FLOW_CDM);
+        }
+      }, 100)
+    }
+
+    return this.element!;
+  }
+
+  _makePropsProxy(props: any): any {
+    const self = this;
+
+    return new Proxy(props as unknown as object, {
+      get(target: Record<string, unknown>, prop: string) {
+        const value = target[prop];
+        return typeof value === 'function' ? value.bind(target) : value;
+      },
+      set(target: Record<string, unknown>, prop: string, value: unknown) {
+        target[prop] = value;
+        self.eventBus().emit(Block.EVENTS.FLOW_CDU, { ...target }, target);
```

Comment:
Можно использовать стрелочные функции, тогда не придеться прибегать к замыканию
```suggestion
      set: (target: Record<string, unknown>, prop: string, value: unknown) => {
        target[prop] = value;
        this.eventBus().emit(Block.EVENTS.FLOW_CDU, { ...target }, target);
```